include/swift/AST/AccessScope.h

```diff
@@ -0,0 +1,91 @@
+//===--- AccessScope.h - Swift Access Scope ---------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_ACCESSSCOPE_H
+#define SWIFT_ACCESSSCOPE_H
+
+#include "swift/AST/AttrKind.h"
+#include "swift/AST/DeclContext.h"
+#include "swift/AST/Module.h"
+#include "llvm/ADT/PointerIntPair.h"
+
+namespace swift {
+
+/// The wrapper around the outermost DeclContext from which
+/// a particular declaration can be accessed.
+class AccessScope {
+  /// The declaration context (if not public) along with a bit saying
+  /// whether this scope is private (or not).
+  llvm::PointerIntPair<const DeclContext *, 1, bool> Value;
+public:
+  AccessScope(const DeclContext *DC, bool isPrivate = false)
+    : Value(DC, isPrivate) {
+        if (!DC || isa<ModuleDecl>(DC))
+          assert(!isPrivate && "public or internal scope can't be private");
+      }
+
+  static AccessScope getPublic() { return AccessScope(nullptr); }
+
+  /// Returns nullptr if access scope is public.
+  const DeclContext *getDeclContext() const { return Value.getPointer(); }
+
+  bool operator==(AccessScope RHS) const { return Value == RHS.Value; }
+  bool operator!=(AccessScope RHS) const { return !(*this == RHS); }
+  bool hasEqualDeclContextWith(AccessScope RHS) const {
+    return getDeclContext() == RHS.getDeclContext();
+  }
+
+  bool isPublic() const { return !Value.getPointer(); }
+  bool isPrivate() const { return Value.getInt(); }
+  bool isFileScope() const {
+    auto DC = getDeclContext();
+    return DC && isa<FileUnit>(DC);
+  }
+
+  /// Returns true if this is a child scope of the specified other access scope.
+  ///
+  /// \see DeclContext::isChildContextOf
+  bool isChildOf(AccessScope AS) const {
```

Comment:
I believe DeclContext::isChildContextOf returns `false` if the two are the same, which implies that `public->isChildOf(public)` should also return false here.